File: 337-house-robber-iii/337-house-robber-iii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    //top-down recursion + memo
    Map<TreeNode, Integer> memo = new HashMap<>();
    public int rob(TreeNode root) {
        //base case
        if(root == null) return 0;
        //check memo
        if(memo.containsKey(root)) return memo.get(root);
        int rob = root.val;
        if(root.left != null) rob += rob(root.left.left) + rob(root.left.right);
        if(root.right != null) rob += rob(root.right.left) + rob(root.right.right);
        int not_rob = rob(root.left) + rob(root.right);
        int res = Math.max(rob, not_rob);
        memo.put(root, res);
        return res;
    }
}

File: 337-house-robber-iii/337-house-robber-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        cache = {}
        def dfs(root):
            nonlocal cache
            if not root:
                return 0 
            if root in cache:
                return cache[root]
            sum = 0
            sum += root.val
            left = dfs(root.left)
            right = dfs(root.right)
            if root.left:
                sum += dfs(root.left.left) + dfs(root.left.right)
            if root.right:
                sum += dfs(root.right.left) + dfs(root.right.right)
            
            cache[root] = max(sum, left + right)
            return cache[root]
            
        return max(dfs(root), dfs(root.left) + dfs(root.right))
            